pkg/storage/gcs/gcp_labels_tags.go

```diff
@@ -35,3 +74,216 @@ func getUserLabels(infraLister configlisters.InfrastructureLister) (map[string]s
+// the passed values for limit, burst(or token bucket) size. If opted for emptyBucket
+// all initial tokens are reserved for the first burst.
+func newLimiter(limit, burst int, emptyBucket bool) *rate.Limiter {
+	limiter := rate.NewLimiter(rate.Every(time.Second/time.Duration(limit)), burst)
+
+	if emptyBucket {
+		limiter.AllowN(time.Now(), burst)
+	}
+
+	return limiter
+}
+
+// toTagValueList converts the tags to an array containing tagValues
+// NamespacedNames.
+func toTagValueList(tags []configv1.GCPResourceTag) []string {
+	if len(tags) <= 0 {
+		return nil
+	}
+
+	list := make([]string, 0, len(tags))
+	for _, tag := range tags {
+		t := fmt.Sprintf("%s/%s/%s", tag.ParentID, tag.Key, tag.Value)
+		list = append(list, t)
+	}
+	return list
+}
+
+// getInfraResourceTagsList returns the user-defined tags present in the
+// status sub-resource of Infrastructure.
+func getInfraResourceTagsList(platformStatus *configv1.PlatformStatus) []configv1.GCPResourceTag {
+	if platformStatus != nil && platformStatus.GCP != nil && platformStatus.GCP.ResourceTags != nil {
+		return platformStatus.GCP.ResourceTags
+	}
+	klog.V(1).Infof("getInfraResourceTagsList: user-defined tag list is empty")
+	return nil
+}
+
+// getTagsList returns the list of tags to apply on the resources.
+func getTagsList(platformStatus *configv1.PlatformStatus) []string {
+	return toTagValueList(getInfraResourceTagsList(platformStatus))
+}
+
+// getFilteredTagList returns the list of tags to apply on the resources after
+// filtering the tags already existing on a given resource.
+func getFilteredTagList(ctx context.Context, platformStatus *configv1.PlatformStatus, client *rscmgr.TagBindingsClient, parent string) []string {
+	return filterTagList(ctx, client, parent, getTagsList(platformStatus))
+}
+
+// filterTagList returns the filtered list of tags to apply on the resources.
+func filterTagList(ctx context.Context, client *rscmgr.TagBindingsClient, parent string, tagList []string) []string {
+	dupTags := make(map[string]bool, len(tagList))
+	for _, k := range tagList {
+		dupTags[k] = false
+	}
+
+	listBindingsReq := &rscmgrpb.ListEffectiveTagsRequest{
+		Parent: parent,
+	}
+	bindings := client.ListEffectiveTags(ctx, listBindingsReq)
+	// a resource can have a maximum of {gcpMaxTagsPerResource} tags attached to it.
+	// Will iterate for {gcpMaxTagsPerResource} times in the worst case scenario, if
+	// none of the break conditions are met. Should the {gcpMaxTagsPerResource} be
+	// increased in future, it should not create an issue, since this is an optimization
+	// attempt to reduce the number the tag write calls by skipping already existing tags,
+	// since it has a quota restriction.
+	for i := 0; i < gcpMaxTagsPerResource; i++ {
+		binding, err := bindings.Next()
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil || binding == nil {
+			klog.V(1).Infof("failed to list effective tags on the %s bucket: %v: %v", parent, binding, err)
```

Comment:
Why don't return this error to the caller so that it can retry the operation?

Reply:
We are trying to save the API calls by finding the already existing tags, I feel retry would not be required, because the main objective would be achieved even without this. Please let me what do you think?

Reply:
Perhaps this should be logged with `klog.V(4).Infof` instead then, since it seems to be more of a debug log?

Reply:
Yeah, but since we are not failing the operation and if it's sporadic/isolated failure, this log would be helpful I think. Could we use `V(1)`, since its logged only during error scenario?

Reply:
We probably could. I'm concerned about this happening and customers seeing it in their logs and mistaking it for something more serious than it is, that's why I suggested we change the level.
Do you think we absolutely need this on `V(1)`? If so, we probably want to rephrase the message so that there would be no confusion.

Reply:
I think `"filterTagList: listing effective tags on the %s bucket ended: %v: %v"` would be helpful for us to identifying if the scenario happened. Please let me WDYT.

Reply:
Alright I won't keep pushing, let's leave it as is

Reply:
suggestion incorporated.